docs(docs/content/docs): add configmap doc (#2940)
openservicemesh/osm: docs/content/docs/osm_config_map.md

type: docs
---

> ## OSM ConfigMap

Reviewer: Nit: see https://deploy-preview-2940--osm-docs.netlify.app/docs/osm_config_map/
```suggestion
# OSM ConfigMap
```

OSM deploys a ConfigMap `osm-config` as a part of its control plane (in the same namespace as that of the osm-controller pod) which can be updated by the mesh owner/operator at any time. The purpose of this ConfigMap is to provide the mesh owner/operator the ability to update some of the mesh configurations based on their needs. The OSM ConfigMap can be found under [charts/osm/templates](https://github.com/openservicemesh/osm/blob/main/charts/osm/templates/osm-configmap.yaml).
To view your `osm-config` in CLI use the `kubectl get` command.
```bash
# Replace osm-system with osm-controller's namespace if using a non-default namespace
kubectl get configmap osm-config -n osm-system -o yaml
```

> ### ConfigMap Values

Reviewer: based on prev suggestion:
```suggestion
## ConfigMap Values
```

| prometheus_scraping | OpenServiceMesh.enablePrometheusScraping | bool | true, false | `"true"` | Enables Prometheus metrics scraping on sidecar proxies. |
| service_cert_validity_duration | OpenServiceMesh.serviceCertValidityDuration | string | 24h, 1h30m (any time duration) | `"24h"` | Sets the service certificate validity duration, represented as a sequence of decimal numbers each with optional fraction and a unit suffix. |
| tracing_enable | OpenServiceMesh.tracing.enable | bool | true, false | `"false"` | Enables Jaeger tracing for the mesh. |
| tracing_address | OpenServiceMesh.tracing.address | string | jaeger.mesh-namespace.svc.cluster.local | `jaeger.osm-system.svc.cluster.local` | Address of the Jaeger deployment, if tracing is enabled. |
| tracing_endpoint | OpenServiceMesh.tracing.endpoint | string | /api/v2/spans | /api/v2/spans | Endpoint for tracing data, if tracing enabled. |
| tracing_port| OpenServiceMesh.tracing.port | int | any non-zero integer value | `"9411"` | Port on which tracing is enabled. |
| use_https_ingress | OpenServiceMesh.useHTTPSIngress | bool | true, false | `"false"`| Enables HTTPS ingress on the mesh. |

> ### Configure OSM ConfigMap

Reviewer: ```suggestion
## Configure OSM ConfigMap
```

### OSM Mesh Upgrade Command
To configure values in `osm-config` use the `osm mesh upgrade` command, so that values changed in the ConfigMap are preserved. See [here](https://github.com/openservicemesh/osm/blob/main/docs/content/docs/upgrade_guide.md) for additional details on `osm mesh upgrade`.

> ```bash
> # Replace osm-system with osm-controller's namespace if using a non-default namespace
> osm mesh upgrade --osm-namespace osm-system --use_https_ingress=true
> ```
> > The `osm-namespace` flag can be omitted if using the default namespace.

Reviewer: this is seeming kind of circular/repetitive since there is a comment before and note after that get at the samething. `osm-system` is the default ns, so why provide it in the example command and then tell folks to remove it? A couple options...
```suggestion
```bash
osm mesh upgrade --use_https_ingress=true [--osm-namespace <namespace>]
```
> The `osm-namespace` flag can be omitted if using the default namespace.
```

Reviewer: OR 
```suggestion
```bash
# Include --osm-namespace flag if using a non-default namespace
osm mesh upgrade --use_https_ingress=true
```
```

Reviewer: or some combination of the two? I'm not sure how we've shown examples of optional flags in other docs but I've seen it done the [ ] way in other repos

`osm mesh upgrade` uses a pre-determined set of flags. Use the `--help` flag to get a list of flags and brief explanation of each flag.
```bash
osm mesh upgrade --help
```
 `Error: unknown flag: --<flag-name>` will occur if an incorrect flag is used.
If an invalid value is used for a flag, an error will occur that explains why the value is invalid.
Example setting boolean flag to an int:
```bash
osm mesh upgrade --enable-egress=3
Error: invalid argument "3" for "--enable-egress" flag: strconv.ParseBool: parsing "3": invalid syntax
```

> ### OSM Mesh Upgrade ConfigMap Flags

Reviewer: seems like this should be nested under the Mesh Upgrade title?
```suggestion
#### OSM Mesh Upgrade ConfigMap Flags
```

```bash
# Replace osm-system with osm-controller's namespace if using a non-default namespace
kubectl patch ConfigMap osm-config -n osm-system -p '{"data":{"use_https_ingress":"true"}}' --type=merge
```
A list of all the keys that can be changed in the ConfigMap can be found in the table [above](#ConfigMap-Values) or in the [`osm-config` yaml](https://github.com/openservicemesh/osm/blob/main/charts/osm/templates/osm-configmap.yaml).

If an incorrect key is used, a new key-value pair will be added into `osm-config`.

If an incorrect value is used, the validating webhook will prevent the change with an error message explaining why the value is invalid.
The below command shows what happens if we patch `use_https_ingress` to a non-boolean value.
```bash
kubectl patch ConfigMap osm-config -n osm-system -p '{"data":{"use_https_ingress":"3"}}' --type=merge
# Validating webhook will deny the change
Error from server (
use_https_ingress: must be a boolean): admission webhook "osm-config-webhook.k8s.io" denied the request:
use_https_ingress: must be a boolean
```

> ### Kubectl Patch Command for Each Key Type

Reviewer: ```suggestion
#### Kubectl Patch Command for Each Key Type
```

| service_cert_validity_duration | string | `"24h"` | `kubectl patch ConfigMap osm-config -n osm-system -p '{"data":{"service_cert_validity_duration":"2m"}}' --type=merge` |
| tracing_address | string | `jaeger.osm-system.svc.cluster.local` | `kubectl patch ConfigMap osm-config -n osm-system -p '{"data":{"tracing_address":"1.2a.b.c3"}}' --type=merge` |
| tracing_endpoint | string | /api/v2/spans | `kubectl patch ConfigMap osm-config -n osm-system -p '{"data":{"tracing_endpoint":"/abracadabra"}}' --type=merge` |
| tracing_port| int | `"9411"` | `kubectl patch ConfigMap osm-config -n osm-system -p '{"data":{"tracing_port":"1234"}}' --type=merge` |

## Validating Webhook

The validating webhook validates changes made to the default fields in the `osm-config` . It prevents users who are updating configurable field values with `kubectl` commands from using values beyond the tested limits. Default fields of the osm-config are the keys in the table [above](#ConfigMap-Values) and can also be found in the [`osm-config` yaml](https://github.com/openservicemesh/osm/blob/main/charts/osm/templates/osm-configmap.yaml).

> The validating webhook yaml can be found [here](https://github.com/openservicemesh/osm/blob/main/charts/osm/templates/validatingwebhook.yaml) and the validation code [here](https://github.com/openservicemesh/osm/blob/main/pkg/configurator/validating_webhook.go). The validating webhook will accept or reject the changed ConfigMap configuration for the value(s) based on the [allowed values](#ConfigMap-Values) for that field(s).

Reviewer: broken link
```suggestion
The validating webhook yaml can be found [here](https://github.com/openservicemesh/osm/blob/main/charts/osm/templates/validatingwebhook.yaml) and the validation code [here](https://github.com/openservicemesh/osm/blob/main/pkg/configurator/validating_webhook.go). The validating webhook will accept or reject the changed ConfigMap configuration for the value(s) based on the [allowed values](#configmap-values) for that field(s).
```

| enable_privileged_init_container| `must be a boolean` |
| envoy_log_level | `invalid log level` |
| outbound_ip_range_exclusion_list | `must be a list of valid IP addresses of the form a.b.c.d/x` |
| permissive_traffic_policy_mode | `must be a boolean` |
| prometheus_scraping | `must be a boolean` |
| service_cert_validity_duration | `invalid time format must be a sequence of decimal numbers each with optional fraction and a unit suffix` |
| tracing_enable | `must be a boolean` |
| tracing_port| <ul><li>`must be an integer`</li><li>`must be between 0 and 65535`</li></ul> |
| use_https_ingress | `must be a boolean` |

> Any changes to the OSM ConfigMap metadata will be rejected with `cannot change metadata`.

### List of Default Fields in ConfigMap
- egress
- enable_debug_server
- enable_privileged_init_container
- envoy_log_level
- permissive_traffic_policy_mode
- prometheus_scraping
- service_cert_validity_duration
- tracing_enable
- use_https_ingress

> All default fields must be present in `osm-config`, removal of one or more default fields will error with `<field>: must be included as it is a default field`.

# Troubleshooting Guide

> See [here](docs/content/docs/troubleshooting/osm_config_map.md) for ConfigMap troubleshooting.

Reviewer: Are we using relative links or absolute links or docs site links?

Author: let's say docs site links for now

type: docs
---

> Table of Contents

Reviewer: It seems like our website automatically adds a ToC to the right hand side of the page so it may be cleaner to remove this?
https://deploy-preview-2940--osm-docs.netlify.app/docs/osm_config_map/

Author: oops didn't realize, will go ahead and remove it

---

# OSM ConfigMap

> OSM deploys a ConfigMap `osm-config` as a part of its control plane (in the same namespace as that of the osm-controller pod) which can be updated by the mesh owner/operator at any time. The purpose of this ConfigMap is to provide the mesh owner/operator the ability to update some of the mesh configurations based on their needs. The OSM ConfigMap can be found under [charts/osm/templates](https://github.com/openservicemesh/osm/blob/main/charts/osm/templates/osm-configmap.yaml).

Reviewer 2: ```suggestion
OSM deploys a ConfigMap `osm-config` as a part of its control plane (in the same namespace as that of the osm-controller pod) which can be updated by the mesh owner/operator at any time. The purpose of this ConfigMap is to provide the mesh owner/operator the ability to update some of the mesh configurations based on their needs. The OSM ConfigMap can be found under [charts/osm/templates](https://github.com/openservicemesh/osm/blob/release-v0.8/charts/osm/templates/osm-configmap.yaml).
```

| envoy_log_level | `--envoy-log-level` | string | `"error"` |
| outbound_ip_range_exclusion_list | `--outbound-ip-range-exclusion-list` | string | `-`|
| permissive_traffic_policy_mode | `--enable-permissive-traffic-policy` | bool | `"false"` |
| prometheus_scraping |`--enable-prometheus-scraping` | bool | `"true"` |
| service_cert_validity_duration | `--service-cert-validity-duration` | string | `"24h"` |
| tracing_enable | `--enable-tracing` | bool | `"false"` |
| tracing_address | `--tracing-address` | string | `jaeger.osm-system.svc.cluster.local` |
| tracing_endpoint | `--tracing-endpoint` | string | /api/v2/spans |
| tracing_port| `--tracing-port` | int | `"9411"` |
| use_https_ingress | `--use-https-ingress` | bool | `"false"`|

### Kubectl Patch Command
To create temporary changes to `osm-config` that will not be preserved across release upgrades, we can use the `kubectl patch` command.
```bash
# Replace osm-system with osm-controller's namespace if using a non-default namespace
kubectl patch ConfigMap osm-config -n osm-system -p '{"data":{"use_https_ingress":"true"}}' --type=merge
```

> A list of all the keys that can be changed in the ConfigMap can be found in the table [above](#configmap-values) or in the [`osm-config` yaml](https://github.com/openservicemesh/osm/blob/main/charts/osm/templates/osm-configmap.yaml).

Reviewer 2: ```suggestion
A list of all the keys that can be changed in the ConfigMap can be found in the table [above](#configmap-values) or in the [`osm-config` yaml](https://github.com/openservicemesh/osm/blob/release-v0.8/charts/osm/templates/osm-configmap.yaml).
```

If an incorrect key is used, a new key-value pair will be added into `osm-config`.

> If an incorrect value is used, the validating webhook will prevent the change with an error message explaining why the value is invalid.

Reviewer: could link to the validating webhook section if you want!

> The below command shows what happens if we patch `use_https_ingress` to a non-boolean value.

Reviewer: ```suggestion
For example, the below command shows what happens if we patch `use_https_ingress` to a non-boolean value.
```
